ipam/space/space.go

```diff
@@ -37,14 +37,40 @@ func (s *Space) Clear() {
 	s.ours = s.ours[:0]
 }
 
-func (s *Space) Allocate() (bool, address.Address) {
-	if len(s.free) == 0 {
-		return false, 0
+// Walk down the free list calling f() on the in-range portions, until
+// f() returns true or we run out of free space.  Return true iff f() returned true
+func (s *Space) walkFree(rangeStart, rangeEnd address.Address, f func(start, end address.Address) bool) bool {
+	for i := 0; i < len(s.free); i += 2 {
+		start, end := s.free[i], s.free[i+1]
+		if start >= rangeEnd {
+			// all remaining free space is completely after range
+			break
+		}
+		// Restrict this block of free space to be in range
+		if start < rangeStart {
+			start = rangeStart
+		}
+		if end > rangeEnd {
+			end = rangeEnd
+		}
+		if start >= end { // no space left in range
```

Comment:
If `rangeStart==rangeEnd` then your assumption that `start>=end` implies 'current free space chunk is completely before the range' falls down.  It seems more bullet-proof to test this last, then it's obvious that, for instance, `Allocate` can never receive an empty range.